docs/api/StreamSend.md

```diff
@@ -21,15 +21,73 @@ QUIC_STATUS
+
+`ClientSendContext`
+
+The valid handle to an open stream object.
 
 # Return Value
 
 The function returns a [QUIC_STATUS](QUIC_STATUS.md). The app may use `QUIC_FAILED` or `QUIC_SUCCEEDED` to determine if the function failed or succeeded.
 
 # Remarks
 
-**TODO**
+This function is used to queue data on a stream to be sent. The function itself is non-blocking and simply queues the data and returns. The app may pass zero or more buffers of data that will be sent on the stream in the order they are passed. The buffers (both the `QUIC_BUFFER`s and the memory they reference) are "owned" by MsQuic (and must not be modified by the app) until MsQuic indicates the `QUIC_STREAM_EVENT_SEND_COMPLETE` event for the send.
+
+By default, data queued via `StreamSend` is not allowed to be sent in 0-RTT packets, but the app may overwrite this by passing the `QUIC_SEND_FLAG_ALLOW_0_RTT` flag. This flag indicates that the data is acceptable to be sent in a 0-RTT packet, but does not guarantee that data will be sent in 0-RTT. There are several reasons it may not be sent in 0-RTT:
+
+- The 0-RTT keys were not available.
+- The server rejected 0-RTT data for some reason.
+- Too much data was queued and it couldn't all fit.
+- The data was sent, but eventually found to have been lost and retransmitted in a 1-RTT packet.
+
+Some apps may open and send on many different streams at a very high rate. In these scenarios, having to call [StreamStart](StreamStart.md) and [StreamShutdown](StreamShutdown.md) for every stream adds unwanted performance overhead. In order to optimize these scenarios, `StreamStart` supports the `QUIC_SEND_FLAG_START` and `QUIC_SEND_FLAG_FIN` flags, which allows the app to do something like this:
+
+```c
+HQUIC Stream;
+MsQuic->StreamOpen(
+    Connection,
+    QUIC_STREAM_OPEN_FLAG_UNIDIRECTIONAL,
+    AppHandler,
+    AppContext,
+    &Stream);
+MsQuic->StreamSend(
+    Stream,
+    &AppData,
+    1,
+    QUIC_SEND_FLAG_START | QUIC_SEND_FLAG_FIN,
+    AppSendContext);
+```
+
+This example opens a new unidirectional stream, and queues a send that starts the stream, sends some app data and gracefully closes the stream. **Note** for the sake of brevity, error handling and clean up has been omitted.
+
+In some scenarios, the app may know that additional data (possibly on a different stream) will soon be queued after the current call to `StreamSend`. In these cases it may be helpful for the app to pass the `QUIC_SEND_FLAG_DELAY_SEND` flag to hint that MsQuic should wait for more data before flushing the connection-wide send queue. **Note** that anything else on the connection *might* still end up triggering the send to flush. Also, be aware that if something else doesn't end up triggering the flush, then MsQuic doesn't have any special mechanism to force a flush after some period of time.
```

Comment:
"Also, be aware that if something else doesn't end up triggering the flush, then MsQuic doesn't have any special mechanism to force a flush after some period of time."

I think you could make a stronger statement here... something like 

"Data passed via StreamSend with the QUIC_SEND_FLAG_DELAY_SEND flag is not guaranteed to be sent until a subsequent StreamSend call on the same stream is performed without the QUIC_SEND_FLAG_DELAY_SEND flag"

Or something like that.

Reply:
Sounds reasonable. Thanks.

Reply:
Also:

I assume it's legal to call StreamSend with 0 bytes and QUIC_SEND_FLAG_DELAY_SEND unset, in order to guarantee any queued data will be sent? Might be useful to call that out.

Reply:
Yes, this is explicitly supported to mimic a "flush" behavior.